### PR TITLE
Sandbox directory syncer

### PR DESCRIPTION
**Pull request by shamsimam**

### CODE REVIEW

File: scheduler/test/cook/test/mesos/sandbox.clj [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+                    (is (= publish-batch-size batch-size))
+                    (swap! task-id->sandbox-publish-history-atom conj @task-id->sandbox-agent)
+                    (.countDown latch)
+                    (send task-id->sandbox-agent sandbox/clear-agent-state @task-id->sandbox-agent)
+                    (await task-id->sandbox-agent)
+                    (Thread/sleep publish-interval-ms))]
+
+      (let [cancel-fn (sandbox/start-sandbox-publisher
+                        task-id->sandbox-agent db-conn publish-batch-size publish-interval-ms)]
+        (.await latch)
```
Comment (pschorf) [inline scheduler/test/cook/test/mesos/sandbox.clj]: Change this to the the version that takes a timeout

**Comment by pschorf** [comment]
Finished my first pass

File: scheduler/src/cook/mesos/sandbox.clj [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+                    (fn mesos-agent-query-cache-swap-fn [c]
+                      (if (cache/has? c agent-hostname)
+                        (cache/hit c agent-hostname)
+                        (cache/miss c agent-hostname run))))
+          val (cache/lookup cs agent-hostname)]
+      (if val @val @run))
+    (catch Exception e
+      (log/error e "Failed to refresh mesos agent state" {:agent agent-hostname}))))
+
+(defn prepare-sandbox-helpers
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply scheduler/src/cook/mesos/sandbox.clj]: Done. Created a record that stores the agent and the cache.

File: scheduler/src/cook/components.clj [hunk 3]
```diff
@@ -91,7 +90,7 @@
                            mesos-gpu-enabled mesos-leader-path mesos-master mesos-master-hosts mesos-principal
                            mesos-role offer-incubate-time-ms progress rebalancer riemann server-port task-constraints]
                           curator-framework framework-id mesos-datomic mesos-datomic-mult mesos-leadership-atom
-                          mesos-offer-cache mesos-pending-jobs-atom]
+                          mesos-offer-cache mesos-pending-jobs-atom sandbox-helper-fns]
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply scheduler/src/cook/components.clj]: Done :)

File: scheduler/src/cook/mesos/sandbox.clj (hunk 2 above)
[the remark this answers is not in the document]
Comment (pschorf) [reply scheduler/src/cook/mesos/sandbox.clj]: What about `prepare-sandbox-publisher`

[the remark this answers is not in the document]
Comment (shamsimam) [reply scheduler/src/cook/mesos/sandbox.clj]: Done.

**Comment by shamsimam** [comment]
Build failed due to flaky `test_show_progress_message` which is fixed in https://github.com/twosigma/Cook/pull/532

**Comment by shamsimam** [comment]
@pschorf  Build is now green :)

**Pull request merged**